wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/hc/probe/ProbeScheduler.java

```diff
@@ -0,0 +1,203 @@
+  }
+
+  private void schedule(
+      String workspaceId, ProbeFactory probeFactory, Consumer<ProbeResult> probeResultConsumer) {
+    ProbeConfig probeConfig = probeFactory.getProbeConfig();
+    Task task = new Task(probeFactory, probeResultConsumer);
+    // scheduleWithFixedDelay is used in favor of scheduleAtFixedRate because in case of bug amount
+    // of scheduled probes start time of tasks may shift and this may lead to a situation when
+    // another probeConfig is needed immediately after the previous one is finished which doesn't
+    // seem a good thing
+    ScheduledFuture scheduledFuture =
+        probesExecutor.scheduleWithFixedDelay(
+            task,
+            probeConfig.getInitialDelaySeconds(),
+            probeConfig.getPeriodSeconds(),
+            TimeUnit.SECONDS);
+
+    List<ScheduledFuture> workspaceProbes =
+        probesFutures.computeIfPresent(
+            workspaceId,
+            (OldKey, scheduledFutures) -> {
+              if (scheduledFutures == null) {
+                scheduledFutures = new ArrayList<>();
+              }
+              scheduledFutures.add(scheduledFuture);
+              return scheduledFutures;
+            });
+    // check whether workspace probes were cancelled concurrently which led to removal of the value
+    // in the map
+    if (workspaceProbes == null) {
+      scheduledFuture.cancel(true);
+      task.cancel();
+    }
+  }
+
+  private class Task implements Runnable {
+    private final ProbeFactory probeFactory;
+    private final Consumer<ProbeResult> probeResultConsumer;
+    private final ProbeConfig probeConfig;
+
+    private int failures = 0;
+    private int successes = 0;
+    private AtomicBoolean cancelled = new AtomicBoolean(false);
+
+    public Task(ProbeFactory probeFactory, Consumer<ProbeResult> probeResultConsumer) {
+      this.probeFactory = probeFactory;
+      this.probeConfig = probeFactory.getProbeConfig();
+      this.probeResultConsumer = probeResultConsumer;
+    }
+
+    @Override
+    public void run() {
+      if (cancelled.get()) {
+        return;
+      }
+      Probe probe = probeFactory.get();
+      timeouts.schedule(
+          new TimeoutProbeTask(probe), TimeUnit.SECONDS.toMillis(probeConfig.getTimeoutSeconds()));
+      boolean success = probe.probe();
+      if (success) {
+        // current success increases successes count and clears failures count
+        successes++;
+        failures = 0;
+
+        if (successes >= probeConfig.getSuccessThreshold()) {
+          // TODO make them completable futures? Then we should ensure that
+          // consecutive calls won't use executors thread time but rather use common thread
+          // pool to perform processing of passed probe result
+          if (cancelled.get()) {
+            return;
+          }
+          // Health check satisfies probeConfig health conditions
+          probeResultConsumer.accept(
+              new ProbeResult(
```

Comment:
What makes me confused is that `ProbeConfig` and `ProbeResult` are not config and result of `Probe` but result and config of scheduling.
Maybe we can encapsulate such logic in the `Probe#probe` method?
Then it will be able to return `ProbeResult` and you will not have to use `ProbeFactory` to construct `ProbeResult`.
If we can't do it in this way please review these objects names.

Reply:
Probe itself is a one-time object since it needs cancelation in some cases. Because of that ProbeConfig needs to be outside of the Probe. 
As to the ProbeResult it is generated not on every probe result but only if it satisfies success/failure thresholds of a probe configuration. Because of that, it cannot be generated by the probe itself, since probe doesn't know how many times it succeeded/failed.

Reply:
I understand. 
Then please add java doc to `Probe` that it should not be used directly but `ProbeScheduler` should be used instead. Am I  wrong?

Reply:
ok